### PR TITLE
Integrate AI-Powered Flashcard Creation into Course Materials, Add Search Bar and Update Sharing UI

### PR DESCRIPTION
**Pull request by TuanDangNguyen2003**
## Acknowledgments  
A big thank you to @holoisme for his excellent work in integrating AI prompts and adapting them for use in creating flashcard decks. His contributions laid the foundation for this exciting feature.  

## Feature Overview  
1. **AI-Powered Flashcard Creation**:  
   - Seamless integration of AI into course materials.  
   - Users can now generate a deck of flashcards by clicking on a document.  
   - Saves time by automating flashcard creation, making it intuitive and engaging.  

2. **Search Bar Addition**:  
   - Added a search bar to improve navigation and usability.  
   - Allows users to quickly find specific user they want to share their deck/card with

3. **Updated Sharing UI**:  
   - Redesigned the UI for sharing cards or decks.  
   - Improved user interface for better aesthetics and functionality.  

### Key Highlights:  
- **User Experience**: Simplifies the creation of study materials, making it effortless and engaging.  
- **AI Integration**: Leverages cutting-edge AI technologies to enhance the platform with a trendy and innovative feature.  
- **Automation**: Eliminates the need for manual creation of flashcards by dynamically generating them from document content.  

## Testing and Coverage  
- All related files have been thoroughly tested, ensuring a robust implementation.  
- Test coverage for all relevant files has reached **100%**, guaranteeing reliability and minimizing potential issues.  

## Impact  
This feature significantly enhances the platform's value by:  
1. Offering an interactive, time-saving solution for students and educators.  
2. Aligning with modern trends in AI-driven education.  
3. Improving overall user engagement with an intuitive and powerful tool.  

We are proud to successfully integrate this feature, marking a meaningful milestone in improving user experience.

### CODE REVIEW

File: edweiss-app/components/memento/CardDisplayComponent.tsx [hunk 1]
```diff
@@ -51,7 +51,6 @@ const CardDisplayComponent: ReactComponent<{
                 }} >
 
                 <TText bold mb="sm">Answer:</TText>
-                {/*<TText>{isAnswerVisible ? card.answer : 'Click to reveal the answer'}</TText>*/}
```
Comment (FlorianDinant) [inline edweiss-app/components/memento/CardDisplayComponent.tsx:54]: You maybe can delete this commented line

File: edweiss-firebase/functions/src/functions/memento/createDeckFromMaterial.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+	const deckCollection = CollectionOf<Memento.Deck>(`users/${userId}/courses/${args.courseId}/decks`);
+
+	let pdfContentCache = "";
+	let step = "0";
+
+	try {
+		// if (doc.type == "slides") {
+		if (true) {
+			// const [rawData] = await admin.storage().bucket().file(doc.url).download();
```
Comment (FlorianDinant) [inline edweiss-firebase/functions/src/functions/memento/createDeckFromMaterial.ts:28]: there is multiple commented lines that you can remove here

File: edweiss-firebase/functions/src/functions/memento/createDeckFromMaterial.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+// 	const question = flashcardElements[i];
+// 	if (i == flashcardElements.length - 1)
+// 		continue;
+// 	const answer = flashcardElements[i + 1];
+
+// 	cards.push({
+// 		question,
+// 		answer
+// 	});
+// }
```
Comment (FlorianDinant) [inline edweiss-firebase/functions/src/functions/memento/createDeckFromMaterial.ts:154]: Please remove those commented lines before merging

File: edweiss-firebase/functions/src/index.ts [hunk 4]
```diff
@@ -5,6 +5,10 @@ admin.initializeApp();
 
 import { FunctionImplementation } from 'utils/firebase';
 
+// const res = admin.storage().bucket().get({
+// 	autoCreate: false
+// });
+
```
Comment (FlorianDinant) [inline edweiss-firebase/functions/src/index.ts:11]: you can delete these comments

**Review by FlorianDinant — APPROVED** [review]
Hello, I have not that much to say, your code is well structured, detailed and really useful to the app! Congratulations! Please try to remove the commented lines that I have commented on before merging.

File: edweiss-app/components/memento/CardDisplayComponent.tsx (hunk 1 above)
Comment (TuanDangNguyen2003) [reply edweiss-app/components/memento/CardDisplayComponent.tsx:54]: Ah yes

File: edweiss-firebase/functions/src/functions/memento/createDeckFromMaterial.ts (hunk 3 above)
Comment (TuanDangNguyen2003) [reply edweiss-firebase/functions/src/functions/memento/createDeckFromMaterial.ts:154]: Ooppss, okie xD

**Pull request merged**